src/main/scala/com/gu/contentapi/utils/LogProcessor.scala

```diff
@@ -0,0 +1,77 @@
+package com.gu.contentapi.utils
+
+import com.amazonaws.services.s3.event.S3EventNotification.{ S3Entity, S3EventNotificationRecord }
+import com.amazonaws.services.s3.model.S3Object
+import com.amazonaws.util.IOUtils
+import com.gu.contentapi.models.{ AcastLog, FastlyLog }
+import com.gu.contentapi.services.S3
+import com.gu.contentapi.Config
+import java.io.InputStream
+import scala.io.Source
+
+sealed abstract class LogProcessor[A] {
+  def process(objects: Iterable[S3EventNotificationRecord]): List[A]
+
+  def download(obj: S3EventNotificationRecord): Option[InputStream] =
+    S3.downloadReport(obj).map(_.getObjectContent)
+}
+
+object LogProcessor {
+  lazy val acast = new AcastLogProcessor()
+  lazy val fastly = new FastlyLogProcessor()
+
+  final class FastlyLogProcessor private[LogProcessor] extends LogProcessor[FastlyLog] {
+    final def process(objects: Iterable[S3EventNotificationRecord]) =
+      normalProcess[FastlyLog](
+        FastlyLog(_),
+        r => FastlyLog.onlyDownloads(r) && FastlyLog.onlyGet(r),
+        objects)
+  }
+
+  final class AcastLogProcessor private[LogProcessor] extends LogProcessor[AcastLog] {
+    final def process(objects: Iterable[S3EventNotificationRecord]) =
+      processAcast(normalProcess[AcastLog](
+        AcastLog(_),
+        r => AcastLog.onlyDownloads(r) && AcastLog.onlySuccessfulReponses(r) && AcastLog.onlyGet(r),
+        objects))
+
+    /**
+     * Logs are aggregated according to three pieces of information: the user agent, the ip address
+     * and the episode url: this is as close as we can get to a unique. For each group, we sum the
+     * number of bytes being sent back and filter out entries that do not go beyond
+     * `Config.minDownloadSize` bytes.
+     */
+    final def processAcast[A](logs: List[AcastLog]): List[AcastLog] =
+      logs.foldLeft(Map.empty[(String, String, String), (AcastLog, Long)]) {
+        case (entries, log) =>
+          val size = log.bytes
+          val key = (log.ipAddress, log.userAgent, log.url)
+          entries + entries.get(key).foldLeft(key -> (log, size)) {
+            case ((key, (log, size)), entry) =>
+              key -> (log, size + entry._2)
+          }
+      }
+        .values
+        .filter(_._2 >= Config.minDownloadSize)
```

Comment:
I think this would be a bit more readable with `collect`:
```
.values
.collect { case (log, size) if size >= Config.minDownloadSize => log }
.toList
```

Reply:
Lovely 😍 I always forget those partial functions, thanks!